### PR TITLE
git blame: persist blame visibility state

### PR DESCRIPTION
**Pull request by taras-yemets**
Closes https://github.com/sourcegraph/sourcegraph/issues/41496
Persist git blame visibility state.
[Loom](https://loom.com/share/1a3d390970b6474abfaa6f00cbae7441)

## Test plan
- CI passes
- Manually tested that git blame works in both Monaco and CodeMirror editors

## App preview:

- [Web](https://sg-web-taras-yemets-persist-blame.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-hgsxtqlewz.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx [hunk 1]
```diff
@@ -204,7 +204,6 @@ class BlameDecorationManager implements PluginValue {
  */
 export const showGitBlameDecorations = Facet.define<BlameHunk[], BlameHunk[]>({
     combine: decorations => decorations.flat(),
-    compareInput: (a, b) => a === b || (a.length === 0 && b.length === 0),
```
Comment (taras-yemets) [inline client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: `compareInput` default comparison rule is `===`. No need to make and extra check for length here.

**Comment by taras-yemets** [comment]
@jjinnii, what do you think about the blame column UX when loading the blame decorations (e.g. opening another file or reloading the tab with the current one with the open blame column - see Loom in the description)? Shall we add any loading indicators or handle the loading state differently?

Current state:
- Monaco: if the blame column is visible (blame button pressed) and no decorations are available yet, we show an empty blame column; when decorations are loaded, we populate the column with data; no horizontal content jumps.
- CoreMirror: if the blame column is visible, the blame column appears at first empty (we can notice it sometimes depends on how long data loading takes), then populated with data; there are horizontal content shifts when the blame column appears.

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx (hunk 1 above)
Comment (philipp-spiess) [reply client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: It's also invalid this way, right? i.e `[a, b]` `[b, c]` both have the same length 🤷

**Review by philipp-spiess — APPROVED** [review]
LGTM but we should probably use `useLocalStorage`, WDYT?

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx (hunk 1 above)
Comment (taras-yemets) [reply client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: Agree, Philipp!
Actually, once the blame decorations are loaded, we don't expect them to change (unless [one of the dependencies](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@fbc955bcea743cdb45dbd0cb4059c5fc705d7696/-/blob/client/web/src/repo/blame/useBlameHunks.ts?L105-114) changes but then the new editor instance would be loaded).

File: client/web/src/repo/blame/useBlameVisibility.tsx [hunk 2]
```diff
@@ -2,8 +2,17 @@ import { BehaviorSubject } from 'rxjs'
...
 
-const isBlameVisible = new BehaviorSubject<boolean>(false)
-const setIsBlameVisible = (isVisible: boolean): void => isBlameVisible.next(isVisible)
+const IS_BLAME_VISIBLE_STORAGE_KEY = 'GitBlame.isVisible'
+const isBlameVisible = new BehaviorSubject<boolean>(!!window.localStorage.getItem(IS_BLAME_VISIBLE_STORAGE_KEY))
+const setIsBlameVisible = (isVisible: boolean): void => {
+    isBlameVisible.next(isVisible)
+
+    if (isVisible) {
+        window.localStorage.setItem(IS_BLAME_VISIBLE_STORAGE_KEY, 'true')
```
[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/repo/blame/useBlameVisibility.tsx]: The `useLocalStorage` hook in its traditional usage doesn't fit this use case as it's not 'reactive' (not sure if it's the right term). I mean that setting some value in the toggle blame button component won't trigger blame column component depending on this value rerendering.
But we can still use the `useLocalStorage` hook logic to handle the logic of the local storage updates but keep the current value as an observable. Example implementation: https://github.com/sourcegraph/sourcegraph/pull/41571/commits/cffc72b54184d55b934a1ebaac22bb729e22fb35. What do you think?

[the remark this answers is not in the document]
Comment (taras-yemets) [reply client/web/src/repo/blame/useBlameVisibility.tsx]: > The useLocalStorage hook in its traditional usage doesn't fit this use case as it's not 'reactive' (not sure if it's the right term). I mean that setting some value in the toggle blame button component won't trigger blame column component depending on this value rerendering.

Unless I'm missing something obvious here...

[the remark this answers is not in the document]
Comment (philipp-spiess) [reply client/web/src/repo/blame/useBlameVisibility.tsx]: > I mean that setting some value in the toggle blame button component won't trigger blame column component depending on this value rerendering.

Wow this is unexpected 😱  If we have two usages of the hook with the same key on the web site I expect both to upgrade. Let's go ahead with whatever workaround you find best then, yeah :)

**Comment by taras-yemets** [comment]
> CoreMirror: if the blame column is visible, the blame column appears at first empty (we can notice it sometimes depends on how long data loading takes), then populated with data; there are horizontal content shifts when the blame column appears 

We can see it in the Loom (PR description) from 00:32.

@fkling, I tried a few approaches to make the blame column visible on the CodeMirror initial render, but unfortunately, it didn't work for me. CodeMirror editor has specific APIs, and maybe I'm missing something obvious in the docs. Can you please help me find the right direction?

File: client/web/src/repo/blob/codemirror/blame-decorations.tsx (hunk 1 above)
Comment (fkling) [reply client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: > It's also invalid this way, right?

It's verifying that the arrays are empty, not that they have the same length. So given your example the check would return `false` (i.e. they are different).

> Actually, once the blame decorations are loaded, we don't expect them to change.

If I remember correctly that didn't use to be the case. If it's the case now all the better 😃

Comment (taras-yemets) [reply client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: > It's verifying that the arrays are empty, not that they have the same length. So given your example the check would return false (i.e. they are different)

Makes sense! Thank you for the clarification!

Comment (philipp-spiess) [reply client/web/src/repo/blob/codemirror/blame-decorations.tsx:207]: > It's verifying that the arrays are empty, not that they have the same length. 

Duh, makes sense yeah. 😅

**Comment by fkling** [comment]
@taras-yemets It looks like the blame data is available *after* the file view was rendered the first time. I think CodeMirror makes gutters only as wide as they have to be (so zero width if there is no data). It might be possible to give gutters a fixed width which would kind of circumvent the problem.

Since the CodeMirror file view is not enabled by default I wouldn't necessarily worry about this in this PR. I'm happy to take a look at it.

**Comment by taras-yemets** [comment]
Thanks a lot for looking into this, @fkling!
I created a follow-up issue for it: https://github.com/sourcegraph/sourcegraph/issues/41791.

**Comment by taras-yemets** [comment]
>It looks like the blame data is available after the file view was rendered the first time. I think CodeMirror makes gutters only as wide as they have to be (so zero width if there is no data). It might be possible to give gutters a fixed width which would kind of circumvent the problem.

@fkling, moving `blameDecorations` [definition](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c7b1d0ffbf73b5d4699340992be77ed970e58fda/-/blob/client/web/src/repo/blob/CodeMirrorBlob.tsx?L122&utm_source=VSCode-2.2.9) outside the component (with hardcoded hunks) doesn't seem to help. It's probably because [`extensions`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c7b1d0ffbf73b5d4699340992be77ed970e58fda/-/blob/client/web/src/repo/blob/CodeMirrorBlob.tsx?L154-180&utm_source=VSCode-2.2.9) point to a new array when `useMemo` dependencies change and `useCodeMirror` hook receives new dependencies. But I'm not sure if that's the case.
Setting gutters fixed width (adding an extra CSS class [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c7b1d0ffbf73b5d4699340992be77ed970e58fda/-/blob/client/web/src/repo/blob/codemirror/blame-decorations.tsx?L166%3A35=&utm_source=VSCode-2.2.9)) don't work for me too 🤷🏻‍♂️ 

Thanks again for looking into it!

**Review by jjinnii — APPROVED** [review]
LGTM. I think the loading state experience with the current fileview (not CodeMirror) is sufficient, especially given we're actively trying to move forward with CodeMirror. I will provide guidance on loading state for CodeMirror in #41791!

**Pull request merged**